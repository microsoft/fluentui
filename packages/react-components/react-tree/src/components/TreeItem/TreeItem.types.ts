import type { ComponentProps, ComponentState, ExtractSlotProps, Slot } from '@fluentui/react-utilities';
import type { ButtonContextValue } from '@fluentui/react-button';
import type { TreeItemContextValue } from '../../contexts';
import { treeItemLevelToken } from '../../utils/tokens';
import * as React from 'react';

export type TreeItemCSSProperties = React.CSSProperties & { [treeItemLevelToken]?: string | number };

export type TreeItemSlots = {
  root: Slot<ExtractSlotProps<Slot<'div'> & { style?: TreeItemCSSProperties }>>;
  content: NonNullable<Slot<'div'>>;
  subtree?: Slot<'span'>;
  /**
   * Expand icon slot,
   * by default renders a chevron icon to indicate opening and closing
   */
  expandIcon?: Slot<'span'>;
  /**
   * Actions slot that renders on the end of tree item
   * when the item is hovered/focused
   */
  actions?: Slot<'span'>;
};

export type TreeItemContextValues = {
  treeItem: TreeItemContextValue;
  button: ButtonContextValue;
};

/**
 * TreeItem Props
 */
export type TreeItemProps = ComponentProps<Partial<TreeItemSlots>> & {
  /**
   * If a TreeItem is a leaf, it'll not present the `expandIcon` slot by default.
   * This attribute is used to force the decision if a TreeItem is a leaf or not. By not providing this property
   * this will be inferred by the presence of a subtree as part of the TreeItem children.
   */
  leaf?: boolean;
};

/**
 * State used in rendering TreeItem
 */
export type TreeItemState = ComponentState<TreeItemSlots> & {
  open: boolean;
  isLeaf: boolean;
  level: number;
  /**
   * By design, a button included on the actions slot should be small
   */
  buttonSize: 'small';
  isActionsVisible: boolean;
};
