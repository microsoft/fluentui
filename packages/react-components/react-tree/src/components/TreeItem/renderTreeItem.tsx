import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TreeItemState, TreeItemSlots, TreeItemContextValues } from './TreeItem.types';
import { TreeItemProvider } from '../../contexts';
import { ButtonContextProvider } from '@fluentui/react-button';

/**
 * Render the final JSX of TreeItem
 */
export const renderTreeItem_unstable = (state: TreeItemState, contextValues: TreeItemContextValues) => {
  const { slots, slotProps } = getSlots<TreeItemSlots>(state);

  return (
    <TreeItemProvider value={contextValues.treeItem}>
      <slots.root {...slotProps.root}>
        <slots.content {...slotProps.content}>
          {slots.expandIcon && <slots.expandIcon {...slotProps.expandIcon} />}
          {slotProps.content.children}
          <ButtonContextProvider value={contextValues.button}>
            {slots.actions && <slots.actions {...slotProps.actions} />}
          </ButtonContextProvider>
        </slots.content>
        {state.open && slots.subtree && <slots.subtree {...slotProps.subtree} />}
      </slots.root>
    </TreeItemProvider>
  );
};
