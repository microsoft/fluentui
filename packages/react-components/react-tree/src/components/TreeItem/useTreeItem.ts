import * as React from 'react';
import { getNativeElementProps, useId, useMergedRefs } from '@fluentui/react-utilities';
import { useEventCallback } from '@fluentui/react-utilities';
import { elementContains } from '@fluentui/react-portal';
import type { TreeItemProps, TreeItemState } from './TreeItem.types';
import { useTreeContext_unstable } from '../../contexts/index';
import { treeDataTypes } from '../../utils/tokens';

/**
 * Create the state required to render TreeItem.
 *
 * The returned state can be modified with hooks such as useTreeItemStyles_unstable,
 * before being passed to renderTreeItem_unstable.
 *
 * @param props - props from this instance of TreeItem
 * @param ref - reference to root HTMLElement of TreeItem
 */
export function useTreeItem_unstable<Value = string>(
  props: TreeItemProps<Value>,
  ref: React.Ref<HTMLDivElement>,
): TreeItemState {
  const contextLevel = useTreeContext_unstable(ctx => ctx.level);

  const id = useId('fui-TreeItem-', props.id);

  const {
    onClick,
    onKeyDown,
    as = 'div',
    value = id,
    itemType = 'leaf',
    'aria-level': level = contextLevel,
    ...rest
  } = props;

  const requestTreeResponse = useTreeContext_unstable(ctx => ctx.requestTreeResponse);

  const [isActionsVisible, setActionsVisible] = React.useState(false);
  const [isAsideVisible, setAsideVisible] = React.useState(true);

  const handleActionsRef = (actions: HTMLDivElement | null) => {
    setAsideVisible(actions === null);
  };

  const open = useTreeContext_unstable(ctx => ctx.openItems.has(value));

  const actionsRef = React.useRef<HTMLDivElement>(null);
  const expandIconRef = React.useRef<HTMLDivElement>(null);
  const layoutRef = React.useRef<HTMLDivElement>(null);
  const subtreeRef = React.useRef<HTMLDivElement>(null);

  const handleClick = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
    onClick?.(event);
    if (event.isDefaultPrevented()) {
      return;
    }
    const isEventFromActions = actionsRef.current && elementContains(actionsRef.current, event.target as Node);
    if (isEventFromActions) {
      return;
    }
    const isEventFromSubtree = subtreeRef.current && elementContains(subtreeRef.current, event.target as Node);
    if (isEventFromSubtree) {
      return;
    }
    const isFromExpandIcon = expandIconRef.current && elementContains(expandIconRef.current, event.target as Node);
    requestTreeResponse({
      event,
      itemType,
      value,
      type: isFromExpandIcon ? treeDataTypes.ExpandIconClick : treeDataTypes.Click,
    });
  });

  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLDivElement>) => {
    onKeyDown?.(event);
    if (event.isDefaultPrevented()) {
      return;
    }
    if (event.currentTarget !== event.target) {
      return;
    }
    switch (event.key) {
      case treeDataTypes.End:
      case treeDataTypes.Home:
      case treeDataTypes.Enter:
      case treeDataTypes.ArrowUp:
      case treeDataTypes.ArrowDown:
      case treeDataTypes.ArrowLeft:
      case treeDataTypes.ArrowRight:
        return requestTreeResponse({ event, itemType, value, type: event.key });
    }
    const isTypeAheadCharacter =
      event.key.length === 1 && event.key.match(/\w/) && !event.altKey && !event.ctrlKey && !event.metaKey;
    if (isTypeAheadCharacter) {
      requestTreeResponse({ event, itemType, value, type: treeDataTypes.TypeAhead });
    }
  });

  const handleActionsVisible = useEventCallback((event: React.FocusEvent | React.MouseEvent) => {
    const isTargetFromSubtree = Boolean(
      subtreeRef.current && elementContains(subtreeRef.current, event.target as Node),
    );
    if (!isTargetFromSubtree) {
      setActionsVisible(true);
    }
  });

  const handleActionsInvisible = useEventCallback((event: React.FocusEvent | React.MouseEvent) => {
    const isTargetFromSubtree = Boolean(
      subtreeRef.current && elementContains(subtreeRef.current, event.target as Node),
    );
    const isRelatedTargetFromActions = Boolean(
      actionsRef.current && elementContains(actionsRef.current, event.relatedTarget as Node),
    );
    if (isRelatedTargetFromActions) {
      return setActionsVisible(true);
    }
    if (!isTargetFromSubtree) {
      return setActionsVisible(false);
    }
  });

  return {
    value,
    open,
    subtreeRef,
    actionsRef: useMergedRefs(actionsRef, handleActionsRef),
    expandIconRef,
    layoutRef,
    itemType,
    isActionsVisible,
    isAsideVisible,
    level,
    components: {
      root: 'div',
    },
    root: getNativeElementProps(as, {
      tabIndex: -1,
      ...rest,
      id,
      ref,
      'aria-level': level,
      'aria-expanded': itemType === 'branch' ? open : undefined,
      role: 'treeitem',
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      onMouseOver: handleActionsVisible,
      onFocus: handleActionsVisible,
      onMouseOut: handleActionsInvisible,
      onBlur: handleActionsInvisible,
    }),
  };
}
