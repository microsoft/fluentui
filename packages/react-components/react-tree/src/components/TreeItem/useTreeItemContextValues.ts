import * as React from 'react';
import type { ButtonContextValue } from '@fluentui/react-button';
import type { TreeItemContextValue } from '../../contexts';
import type { TreeItemContextValues, TreeItemState } from './TreeItem.types';

export function useTreeItemContextValues_unstable(
  state: Pick<TreeItemState, 'buttonSize' | 'isActionsVisible'>,
): TreeItemContextValues {
  const { buttonSize, isActionsVisible } = state;

  const treeItem = React.useMemo<TreeItemContextValue>(() => ({ isActionsVisible }), [isActionsVisible]);
  const button = React.useMemo<ButtonContextValue>(() => ({ size: buttonSize }), [buttonSize]);

  return { treeItem, button };
}
