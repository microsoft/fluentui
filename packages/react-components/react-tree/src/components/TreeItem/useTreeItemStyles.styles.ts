import { GriffelStyle, makeResetStyles, makeStyles, mergeClasses } from '@griffel/react';
import type { TreeItemCSSProperties, TreeItemSlots, TreeItemState } from './TreeItem.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import { treeItemLevelToken } from '../../utils/tokens';
import { treeItemLayoutClassNames } from '../TreeItemLayout/useTreeItemLayoutStyles.styles';
import { treeItemPersonaLayoutClassNames } from '../TreeItemPersonaLayout/useTreeItemPersonaLayoutStyles.styles';

export const treeItemClassNames: SlotClassNames<TreeItemSlots> = {
  root: 'fui-TreeItem',
};

const useBaseStyles = makeResetStyles({
  position: 'relative',
  cursor: 'pointer',
  display: 'flex',
  flexDirection: 'column',
  boxSizing: 'border-box',
  backgroundColor: tokens.colorSubtleBackground,
  color: tokens.colorNeutralForeground2,
  paddingRight: tokens.spacingHorizontalNone,
  // if using createCustomFocusIndicatorStyle then we need to remove default outline styles provided by the browser
  ':focus': {
    outlineStyle: 'none',
  },
  ':focus-visible': {
    outlineStyle: 'none',
  },
  // This adds the focus outline for the TreeItemLayout element
  ...createCustomFocusIndicatorStyle(
    {
      borderRadius: tokens.borderRadiusMedium,
      outlineColor: tokens.colorStrokeFocus2,
      outlineRadius: tokens.borderRadiusMedium,
      // FIXME: tokens.strokeWidthThick causes some weird bugs
      outlineWidth: '2px',
      outlineStyle: 'solid',
    },
    {
      customizeSelector: selector =>
        `${selector} > .${treeItemLayoutClassNames.root}, ${selector} > .${treeItemPersonaLayoutClassNames.root}`,
    },
  ),
});

type StaticLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
type StaticLevelProperty = `level${StaticLevel}`;

const useStyles = makeStyles({
  ...(Object.fromEntries(
    Array.from<never, [StaticLevelProperty, TreeItemCSSProperties]>({ length: 10 }, (_, index) => [
      `level${(index + 1) as StaticLevel}`,
      { [treeItemLevelToken]: index + 1 },
    ]),
  ) as Record<StaticLevelProperty, GriffelStyle>),
});

/**
 * Apply styling to the TreeItem slots based on the state
 */
export const useTreeItemStyles_unstable = (state: TreeItemState): TreeItemState => {
  const baseStyles = useBaseStyles();
  const styles = useStyles();

  const { level } = state;

  state.root.className = mergeClasses(
    treeItemClassNames.root,
    baseStyles,
    isStaticallyDefinedLevel(level) && styles[`level${level}` as StaticLevelProperty],
    state.root.className,
  );

  return state;
};

function isStaticallyDefinedLevel(level: number): level is StaticLevel {
  return level >= 1 && level <= 10;
}
