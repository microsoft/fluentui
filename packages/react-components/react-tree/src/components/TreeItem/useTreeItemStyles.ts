import { GriffelStyle, makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TreeItemCSSProperties, TreeItemSlots, TreeItemState } from './TreeItem.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';
import { createFocusOutlineStyle } from '@fluentui/react-tabster';
import { useTreeContext_unstable } from '../../contexts/index';
import { treeItemLevelToken } from '../../utils/tokens';

export const treeItemClassNames: SlotClassNames<TreeItemSlots> = {
  root: 'fui-TreeItem',
  content: 'fui-TreeItem__content',
  subtree: 'fui-TreeItem__subtree',
  expandIcon: 'fui-TreeItem__expandIcon',
  actions: 'fui-TreeItem__actions',
};

type StaticLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
type StaticLevelProperty = `level${StaticLevel}`;

const useRootStyles = makeStyles({
  ...(Object.fromEntries(
    Array.from<never, [StaticLevelProperty, TreeItemCSSProperties]>({ length: 10 }, (_, index) => [
      `level${(index + 1) as StaticLevel}`,
      { [treeItemLevelToken]: index + 1 },
    ]),
  ) as Record<StaticLevelProperty, GriffelStyle>),
});

/**
 * Styles for the content slot
 */
const useContentStyles = makeStyles({
  base: {
    position: 'relative',
    cursor: 'pointer',
    display: 'flex',
    backgroundColor: tokens.colorSubtleBackground,
    color: tokens.colorNeutralForeground2,
    paddingRight: tokens.spacingHorizontalNone,
    paddingLeft: `calc((var(${treeItemLevelToken}, 1) - 1) * ${tokens.spacingHorizontalXXL})`,
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
    ':active': {
      color: tokens.colorNeutralForeground2Pressed,
      backgroundColor: tokens.colorSubtleBackgroundPressed,
      // TODO: stop using treeItemClassNames.expandIcon for styling
      [`& .${treeItemClassNames.expandIcon}`]: {
        color: tokens.colorNeutralForeground3Pressed,
      },
    },
    ':hover': {
      color: tokens.colorNeutralForeground2Hover,
      backgroundColor: tokens.colorSubtleBackgroundHover,
      // TODO: stop using treeItemClassNames.expandIcon  for styling
      [`& .${treeItemClassNames.expandIcon}`]: {
        color: tokens.colorNeutralForeground3Hover,
      },
    },
  },
  focusIndicator: createFocusOutlineStyle(),
  // Appearance variations
  subtle: {},
  'subtle-alpha': {
    ':hover': {
      backgroundColor: tokens.colorSubtleBackgroundLightAlphaHover,
    },
    ':active': {
      backgroundColor: tokens.colorSubtleBackgroundLightAlphaPressed,
    },
  },
  transparent: {
    backgroundColor: tokens.colorTransparentBackground,
    ':hover': {
      backgroundColor: tokens.colorTransparentBackgroundHover,
    },
    ':active': {
      backgroundColor: tokens.colorTransparentBackgroundPressed,
    },
  },
  leaf: {
    paddingLeft: `calc(var(${treeItemLevelToken}, 1) * ${tokens.spacingHorizontalXXL})`,
  },
});

/**
 * Styles for the expand icon slot
 */
const useExpandIconStyles = makeStyles({
  base: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minWidth: '24px',
    boxSizing: 'border-box',
    color: tokens.colorNeutralForeground3,
    ...shorthands.padding(tokens.spacingVerticalXS, 0),
  },
});

/**
 * Styles for the action icon slot
 */
const useActionsStyles = makeStyles({
  base: {
    display: 'flex',
    opacity: '0',
    position: 'absolute',
    right: 0,
    top: 0,
    marginLeft: 'auto',
    ...shorthands.padding(0, tokens.spacingHorizontalS),
  },
  open: {
    opacity: '1',
    position: 'relative',
  },
});

export const expandIconInlineStyles = {
  90: { transform: `rotate(90deg)` },
  0: { transform: `rotate(0deg)` },
  180: { transform: `rotate(180deg)` },
} as const;

/**
 * Apply styling to the TreeItem slots based on the state
 */
export const useTreeItemStyles_unstable = (state: TreeItemState): TreeItemState => {
  const rootStyles = useRootStyles();
  const contentStyles = useContentStyles();
  const expandIconStyles = useExpandIconStyles();
  const actionsStyles = useActionsStyles();

  const appearance = useTreeContext_unstable(ctx => ctx.appearance);

  const { actions, subtree, expandIcon, isActionsVisible: showActions, level } = state;

  state.root.className = mergeClasses(
    treeItemClassNames.root,
    isStaticallyDefinedLevel(level) && rootStyles[`level${level}` as StaticLevelProperty],
    state.root.className,
  );

  state.content.className = mergeClasses(
    treeItemClassNames.content,
    contentStyles.base,
    contentStyles[appearance],
    contentStyles.focusIndicator,
    state.isLeaf && contentStyles.leaf,
    state.content.className,
  );

  if (expandIcon) {
    expandIcon.className = mergeClasses(treeItemClassNames.expandIcon, expandIconStyles.base, expandIcon.className);
  }
  if (actions) {
    actions.className = mergeClasses(
      treeItemClassNames.actions,
      actionsStyles.base,
      showActions && actionsStyles.open,
      actions.className,
    );
  }
  if (subtree) {
    subtree.className = mergeClasses(treeItemClassNames.subtree, subtree.className);
  }

  return state;
};

function isStaticallyDefinedLevel(level: number): level is StaticLevel {
  return level >= 1 && level <= 10;
}
