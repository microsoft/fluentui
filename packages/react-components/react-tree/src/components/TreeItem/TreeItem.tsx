import * as React from 'react';
import { useTreeItem_unstable } from './useTreeItem';
import { renderTreeItem_unstable } from './renderTreeItem';
import { useTreeItemStyles_unstable } from './useTreeItemStyles';
import type { TreeItemProps } from './TreeItem.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useTreeItemContextValues_unstable } from './useTreeItemContextValues';

/**
 * TreeItem component - TODO: add more docs
 */
export const TreeItem: ForwardRefComponent<TreeItemProps> = React.forwardRef((props, ref) => {
  const state = useTreeItem_unstable(props, ref);

  useTreeItemStyles_unstable(state);
  const contextValues = useTreeItemContextValues_unstable(state);
  return renderTreeItem_unstable(state, contextValues);
});

TreeItem.displayName = 'TreeItem';
