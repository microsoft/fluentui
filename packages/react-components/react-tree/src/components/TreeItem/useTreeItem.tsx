import * as React from 'react';
import { getNativeElementProps, isResolvedShorthand, resolveShorthand, useId } from '@fluentui/react-utilities';
import { ChevronRight12Regular } from '@fluentui/react-icons';
import { useFluent_unstable } from '@fluentui/react-shared-contexts';
import { useEventCallback } from '@fluentui/react-utilities';
import { useFocusableGroup } from '@fluentui/react-tabster';
import { expandIconInlineStyles } from './useTreeItemStyles';
import { ArrowLeft, ArrowRight, Enter } from '@fluentui/keyboard-keys';
import { useMergedRefs } from '@fluentui/react-utilities';
import { elementContains } from '@fluentui/react-portal';
import type { TreeItemProps, TreeItemState } from './TreeItem.types';
import { useTreeContext_unstable } from '../../contexts/index';

/**
 * Create the state required to render TreeItem.
 *
 * The returned state can be modified with hooks such as useTreeItemStyles_unstable,
 * before being passed to renderTreeItem_unstable.
 *
 * @param props - props from this instance of TreeItem
 * @param ref - reference to root HTMLElement of TreeItem
 */
export const useTreeItem_unstable = (props: TreeItemProps, ref: React.Ref<HTMLDivElement>): TreeItemState => {
  const [children, subtreeChildren] = React.Children.toArray(props.children);

  const contextLevel = useTreeContext_unstable(ctx => ctx.level);
  const {
    content,
    subtree,
    expandIcon,
    leaf: isLeaf = subtreeChildren === undefined,
    actions,
    as = 'div',
    onClick,
    onKeyDown,
    ['aria-level']: level = contextLevel,
    ...rest
  } = props;

  const requestOpenChange = useTreeContext_unstable(ctx => ctx.requestOpenChange);
  const focusFirstSubtreeItem = useTreeContext_unstable(ctx => ctx.focusFirstSubtreeItem);
  const focusSubtreeOwnerItem = useTreeContext_unstable(ctx => ctx.focusSubtreeOwnerItem);

  const id = useId('fui-TreeItem-', props.id);

  const isBranch = !isLeaf;

  const open = useTreeContext_unstable(ctx => isBranch && ctx.openItems.has(id));
  const { dir, targetDocument } = useFluent_unstable();
  const expandIconRotation = open ? 90 : dir !== 'rtl' ? 0 : 180;
  const groupperProps = useFocusableGroup();

  const actionsRef = React.useRef<HTMLElement>(null);
  const expandIconRef = React.useRef<HTMLElement>(null);
  const subtreeRef = React.useRef<HTMLElement>(null);

  const handleArrowRight = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (open && isBranch) {
      focusFirstSubtreeItem(event.currentTarget);
    }
    if (isBranch && !open) {
      requestOpenChange({ event, open: true, type: 'arrowRight' });
    }
  };
  const handleArrowLeft = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (isLeaf || !open) {
      focusSubtreeOwnerItem(event.currentTarget);
    }
    if (isBranch && open) {
      requestOpenChange({ event, open: false, type: 'arrowLeft' });
    }
  };
  const handleEnter = (event: React.KeyboardEvent<HTMLDivElement>) => {
    // if Enter keydown event comes from actions, ignore it
    if (actionsRef.current && elementContains(actionsRef.current, event.target as Node)) {
      return;
    }
    if (isBranch) {
      requestOpenChange({ event, open: !open, type: 'enter' });
    }
  };

  const handleClick = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
    onClick?.(event);
    //  if click event originates from actions, ignore it
    if (actionsRef.current && elementContains(actionsRef.current, event.target as Node)) {
      return;
    }
    if (isBranch) {
      const isFromExpandIcon = expandIconRef.current && elementContains(expandIconRef.current, event.target as Node);
      requestOpenChange({ event, open: !open, type: isFromExpandIcon ? 'expandIconClick' : 'click' });
    }
    event.stopPropagation();
  });

  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLDivElement>) => {
    onKeyDown?.(event);
    if (event.isDefaultPrevented()) {
      return;
    }
    switch (event.key) {
      case Enter: {
        return handleEnter(event);
      }
      case ArrowRight: {
        return handleArrowRight(event);
      }
      case ArrowLeft: {
        return handleArrowLeft(event);
      }
    }
  });

  const [isActionsVisible, setActionsVisible] = React.useState(false);
  const showActions = useEventCallback((event: React.SyntheticEvent) => {
    const isInternalEvent = subtreeRef.current?.contains(event.target as Node) ?? false;
    if (!isInternalEvent) {
      setActionsVisible(true);
    }
  });
  const hideActions = useEventCallback((event: React.SyntheticEvent) => {
    const isInternalEvent = subtreeRef.current?.contains(event.target as Node) ?? false;
    if (!isInternalEvent) {
      setActionsVisible(false);
    }
  });

  // Listens to focusout event on the document to ensure treeitem actions visibility on portal scenarios
  // TODO: find a better way to ensure this behavior
  React.useEffect(() => {
    if (actionsRef.current) {
      const handleFocusOut = (event: FocusEvent) => {
        setActionsVisible(elementContains(actionsRef.current, event.relatedTarget as Node));
      };
      targetDocument?.addEventListener('focusout', handleFocusOut, { passive: true });
      return () => {
        targetDocument?.removeEventListener('focusout', handleFocusOut);
      };
    }
  }, [targetDocument]);

  return {
    isLeaf,
    open,
    level,
    buttonSize: 'small',
    isActionsVisible: actions ? isActionsVisible : false,
    components: {
      content: 'div',
      root: 'div',
      expandIcon: 'span',
      actions: 'span',
      subtree: 'span',
    },
    subtree: resolveShorthand(subtree, {
      required: Boolean(subtreeChildren),
      defaultProps: {
        children: subtreeChildren,
        ref: useMergedRefs(subtreeRef, isResolvedShorthand(subtree) ? subtree.ref : undefined),
      },
    }),
    content: resolveShorthand(content, {
      required: true,
      defaultProps: {
        children,
        ...groupperProps,
      },
    }),
    root: getNativeElementProps(as, {
      ...rest,
      id,
      ref,
      tabIndex: 0,
      children: null,
      'aria-level': level,
      // FIXME: tabster fails to navigate when aria-expanded is true
      // 'aria-expanded': isBranch ? isOpen : undefined,
      role: 'treeitem',
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      onMouseOver: actions ? showActions : undefined,
      onFocus: actions ? showActions : undefined,
      onMouseOut: actions ? hideActions : undefined,
      onBlur: actions ? hideActions : undefined,
    }),
    expandIcon: resolveShorthand(expandIcon, {
      required: isBranch,
      defaultProps: {
        children: <ChevronRight12Regular style={expandIconInlineStyles[expandIconRotation]} />,
        'aria-hidden': true,
        ref: useMergedRefs(isResolvedShorthand(expandIcon) ? expandIcon.ref : undefined, expandIconRef),
      },
    }),
    actions: resolveShorthand(actions, {
      defaultProps: {
        ref: useMergedRefs(isResolvedShorthand(actions) ? actions.ref : undefined, actionsRef),
      },
    }),
  };
};
