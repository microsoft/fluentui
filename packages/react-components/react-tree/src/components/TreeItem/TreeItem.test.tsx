import * as React from 'react';
import { render } from '@testing-library/react';
import { TreeItem } from './TreeItem';
import { isConformant } from '../../testing/isConformant';
import { TreeItemProps } from './TreeItem.types';
import { treeItemClassNames } from './useTreeItemStyles';

describe('TreeItem', () => {
  isConformant<TreeItemProps>({
    Component: TreeItem,
    displayName: 'TreeItem',
    // primarySlot: 'groupper',
    getTargetElement(renderResult, attr) {
      return renderResult.container.querySelector(`.${treeItemClassNames.root}`) ?? renderResult.container;
    },
    disabledTests: ['component-has-static-classnames-object'],
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            expandIcon: 'Test Expand Icon',
            iconBefore: 'Test Icon Before',
            iconAfter: 'Test Icon After',
            actions: 'test Actions',
            aside: 'test Aside',
          },
        },
      ],
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TreeItem>Default TreeItem</TreeItem>);
    expect(result.container).toMatchSnapshot();
  });
});
