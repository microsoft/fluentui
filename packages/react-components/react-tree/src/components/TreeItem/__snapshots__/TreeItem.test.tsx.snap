// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeItem renders a default state 1`] = `
<div>
  <div
    aria-level="0"
    class="fui-TreeItem"
    id="fui-TreeItem-7"
    role="treeitem"
    tabindex="0"
  >
    <div
      class="fui-TreeItem__content"
      data-tabster="{\\"groupper\\":{}}"
    >
      Default TreeItem
    </div>
  </div>
</div>
`;
